plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,45 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+
+			dnsErrBufOccured := false
+			dnsErrOverflowOccured := false
+			var perr *dns.Error
+
+			if proto == "udp" {
+				if errors.As(err, &perr) {
+					if errors.Is(err, dns.ErrBuf) {
+						dnsErrBufOccured = true
+					}
+				}
+
+				if strings.Contains(err.Error(), "overflow") {
+					dnsErrOverflowOccured = true
+				}
+			}
+
+			if dnsErrBufOccured || dnsErrOverflowOccured {
+				// Only if response message id matches the request message id -
+				// Copy the state, set TC bit and break.
+				if state.Req.Id == ret.Id {
+					newRet := state.Req.Copy()
```

Comment:
Why is a new return message created by copying the _request_ (`state.Req`) instead of the response (`ret`)?

Can the code instead alter `ret` (i.e. set TC bit, and empty the answer/extra/auth sections?)

Reply:
I was thinking about it on why the copy was made from the original request instead of original response. Not sure if there was any specific reason (which I don't know) why this was done by Patrick in his original PR. I will reach out to Patrick and see if I am missing something.

Reply:
It looks like this was discussed in comments @phealy's original PR.  Original he used the `ret`, but now I see it changed, with the following rationale ...

> ... I think it would actually be cleaner to build a new response object that we guarantee is in a clean state.

I'm not convinced this is better though.  I think it would be simpler to use the original `ret`, set truncation bit, and clear out answer/extra/auth sections.

Reply:
Yes, I have made the suggested changes. Would you mind taking another look please.